proxy/http2/Http2ConnectionState.cc

```diff
@@ -965,9 +965,12 @@ Http2ConnectionState::main_event_handler(int event, void *edata)
   --recursion;
   if (recursion == 0 && ua_session && !ua_session->is_recursing()) {
     if (this->ua_session->ready_to_free()) {
-      this->ua_session->free();
-      // After the free, the Http2ConnectionState object is also freed.
-      // The Http2ConnectionState object is allocted within the Http2ClientSession object
```

Comment:
since we have #3228, this change is no longer needed.

Reply:
Actually, we are not guaranteed to acquire the connection_state->mutex before this function. Then this change is definitely needed.
we are also doing `fini_event = nullptr`, `--recursion`, `++recursion`, ... operations that need locks. But these are not protected. @masaori335 @maskit @shinrich